### ISSUE TITLE
Potential Lookahead Problem 

### ISSUE THREAD

**Issue by cssd0915** [vetludo/Cluster-Quant-Trading#1]
Hello. Thank you for sharing your project, it looks interesting. 

However, I think there are look-ahead problems in your code, and that is the reason why the result looks so good. 

There are 3 places (code block) that has look-ahead problem:

1. under compute_atr and compute_macd method:
```atr.sub(atr.mean()).div(atr.std())``` and ```macd.sub(macd.mean()).div(macd.std())```
To my best understanding, what you are doing here is ```nomalization```, however, for data point at time ```T``` it should not be able to see the ```mean``` that is computed from data point ```T+1``` to the end of the series (lets say data point ```N```).  You can take a look from this blog: [stackexchange](https://stats.stackexchange.com/questions/442739/look-ahead-bias-induced-by-standardization-of-a-time-series)

3. On code block 9: 
```data.loc[:, factors] = data.groupby(level='ticker', group_keys=False)[factors].apply(lambda x: x.fillna(x.mean()))``` 

I think here, you should not ```fillna``` with ```mean```. We can think about when we are computing ```RSI``` with window size is 14, we should not fill the first 14 rows with mean of RSI computed from 15th row and beyond. Because the first 14 rows should be used as an ipnut data to compute ```RSI``` for the 15th row, it does not make sense if we fill the ```mean``` from 15th row to N for the first 14 rows, in real-time situation, the first 14 rows are impossible to know the ```mean``` which is computed by the data in the future. 

Once again, I am thankful that you shared your idea and code, I make some improvement and fix the look-ahead problem, I will upload it in Code Ocean later, if possible please take a look and hopefully we can exchange some idea in the future. 

Update: here is the notebook: [cssd_github](https://github.com/cssd0915/Unsupervised-Quant-Trading/blob/d3d88a45bf3dbe5e96cdb0857fe07f978ea81221/unsupervised_quant_trading.ipynb)

Thank you!

**Comment by vetludo** [comment]
Thank you very much for the correction. You are right that I made a mistake. I will keep it in mind.

**Comment by cssd0915** [comment]
> Thank you very much for the correction. You are right that I made a mistake. I will keep it in mind.

Hi thanks for the quick reply. I noticed that you uploaded some ML assignment (?) . I was wondering if you are also a student, studying at the ML field. Or has your team already started your project?

**Comment by vetludo** [comment]
Yes, you are right. I have worked in Machine Learning field specialized in Finance for over a decade. Recently, I am studying in Master Degree of Machine Learning and AI and doing research project in portfolio optimization.
